tools/2025-03-workflow-changes.md

```diff
@@ -0,0 +1,260 @@
+    
+    ghsetup --> logic
+    devsetup --> logic
+    cloudsetup --> work
+    work --> cloudcleanup
+    logic --> ghcleanup
+    logic --> devcleanup
+
+    classDef io stroke-dasharray: 5 5
+```
+
+_Fig. 2: Testable automation that can run from a GitHub workflow or a dev machine_
+
+#### Workflows can be run on any fork without requiring access to the `radius-project` repos, and without having a Radius GitHub security role.
+
+```mermaid
+flowchart LR
+    fork["Fork the repo"]
+    setupCreds["Set up credentials"]
+    invoke["Invoke the forked workflow"]
+    
+    fork --> setupCreds
+    setupCreds --> invoke
+```
+
+_Fig. 3: Running a GitHub workflow on a fork_
+
+#### Use the GitHub CLI for default security context
+
+On a developer machine, if the dev is logged into the GitHub CLI, the security context is automatically passed to GitHub by the CLI. For any action that is taken in GitHub we should default to using the GitHub CLI so that this context works locally and within workflows.
+
+In some cases we will choose to use a GitHub CLI call instead of a GitHub provided workflow action so that we can have runnable and testable code on a developer machine (because GitHub actions are not locally testable.)
+
+#### Use GitHub actions for setup, cleanup, and matrix operations
+
+The use of actions should be limited to setup and cleanup that is unique to GitHub workflows or runners because any logic that exists in a GitHub action cannot be executed on a developer machine. Some examples are below:
+
+* Cloning repos
+* Installing tools
+* Retrieving stored secrets
+* Publishing results
+
+> NOTE: The above examples are also needed on a developer machine, but the Implementation is different.
+
+:warning: When using GitHub actions that are published on other repositories, we are [placing our trust](https://arstechnica.com/information-technology/2025/03/supply-chain-attack-exposing-credentials-affects-23k-users-of-tj-actions/) in the authors of that repo that they will prevent malicious code from executing. Choose wisely, and consider forking and customizing a GitHub action instead of calling it directly.
+
+#### Reusable logic exists in custom GitHub actions instead of copy/paste to multiple workflows
+
+Apply the DRY principle (don't repeat yourself).
+
+#### Core logic is runnable from Make commands
+
+Any core logic that is currently in GitHub workflows will be moved to Make commands that are runnable on a developer machine (with some environment setup).
+
+Developer automation is already provided through Make. Some of the GitHub workflows also use the same Make commands, e.g. `make build` is invoked from GitHub workflows.
+
+Some of the existing Make commands invoke scripts. This pattern will continue in order to keep all significant logic isolated to scripts because we do not want to move the complexity problem from GitHub workflows into Make. Simple composite commands should remain within the Make files as they are today.
```

Comment:
Any downsides of this approach that we need to consider?

Reply:
Editing Make files isn't what I think of as fun because they can be complex, but putting anything that is moderatly complex into a shell script that is invoked from Make is helpful to isolate the complexity.

The advantage of using Make and shell scripts is that it leverages a pattern we already have in place.